### PR TITLE
feat(ha): add apis for volume republish and targets destroy for switchover 

### PR DESCRIPTION
**Pull request by Abhinandan-Purkait**
1. Add volume republish api.
2. Add volume destroy shutdown targets api.
3. Add changes to the reconcilers to not act on shutdown states.

### CODE REVIEW

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 1]
```diff
@@ -274,19 +274,20 @@ impl ResourcePublishing for OperationGuardArc<VolumeSpec> {
...
-        let operation =
-            VolumeOperation::Publish((nexus_node.clone(), nexus_id.clone(), request.share));
+        let operation = VolumeOperation::Publish((
+            nexus_node.clone(),
+            nexus_id.clone(),
+            request.share,
+            request.republish(),
+        ));
         let spec_clone = self.start_update(registry, &state, operation).await?;
-
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]: keep the newlines to separate the `start_update` and `validate_update_step` from the rest

File: control-plane/agents/src/bin/core/volume/specs.rs [hunk 2]
```diff
@@ -1111,15 +1111,6 @@ pub(crate) async fn get_volume_target_node(
     status: &VolumeState,
     request: &PublishVolume,
 ) -> Result<NodeId, SvcError> {
-    // We can't configure a new target_node if the volume is currently published
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/specs.rs:1114]: we could keep this and check if request is not a republish?

File: control-plane/agents/src/bin/core/volume/specs.rs [hunk 3]
```diff
@@ -1219,16 +1210,20 @@ impl SpecOperationsHelper for VolumeSpec {
...
-                            Err(SvcError::VolumeAlreadyPublished {
-                                vol_id: self.uuid_str(),
-                                node: target.node().to_string(),
-                                protocol: format!("{:?}", target.protocol()),
-                            })
+                            if !republish {
+                                Err(SvcError::VolumeAlreadyPublished {
+                                    vol_id: self.uuid_str(),
+                                    node: target.node().to_string(),
+                                    protocol: format!("{:?}", target.protocol()),
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/specs.rs]: does protocol have a to_string?

File: control-plane/csi-driver/controller/src/client.rs [hunk 4]
```diff
@@ -283,11 +283,12 @@ impl IoEngineApiClient {
         volume_id: &uuid::Uuid,
         node: Option<&str>,
         protocol: VolumeShareProtocol,
+        republish: bool,
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/controller/src/client.rs]: I'm not sure that we should be exposing this via rest to prevent misuse..

File: control-plane/grpc/proto/v1/volume/volume.proto [hunk 5]
```diff
@@ -169,6 +169,8 @@ message PublishVolumeRequest {
   optional string target_node = 2;
   // share protocol
   optional VolumeShareProtocol share = 3;
+  // republish the volume on the node, without removing existing nexus if any
```
Comment (tiagolobocastro) [inline control-plane/grpc/proto/v1/volume/volume.proto]:
```suggestion
  // allows republishing the volume on the node by shutting down the existing target first
  // once the new target is ready you must destroy the previous target
```

Comment (tiagolobocastro) [reply control-plane/grpc/proto/v1/volume/volume.proto]: Actually I'm starting to wonder if we should have a republish message, which fails if the volume is not currently published and maybe returns the old target uuid so that we may provide the cluster agent a uuid to destroy old targets, wdyt?

File: tests/bdd/features/ha/nexus-switchover/nexus_switchover.feature [hunk 6]
```diff
@@ -0,0 +1,10 @@
+Feature: Nexus Switchover IO test
```
Comment (tiagolobocastro) [inline tests/bdd/features/ha/nexus-switchover/nexus_switchover.feature]: Let's replace all nexuses here with target please

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 1 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/operations.rs]: Oops, will add it back.

File: control-plane/agents/src/bin/core/volume/specs.rs (hunk 2 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/specs.rs:1114]: Yeah. Right

File: control-plane/agents/src/bin/core/volume/specs.rs (hunk 3 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/specs.rs]: This was exsisting code, and seems like we need to change the SvcError's parameter. Will take this up as refactor PR.

File: control-plane/csi-driver/controller/src/client.rs (hunk 4 above)
Comment (Abhinandan-Purkait) [reply control-plane/csi-driver/controller/src/client.rs]: Yeah, I also had this thought but I added it so that we can have the bdd-tests, as the bdd tests use the rest api.

File: control-plane/grpc/proto/v1/volume/volume.proto (hunk 5 above)
Comment (Abhinandan-Purkait) [reply control-plane/grpc/proto/v1/volume/volume.proto]: That's a very good suggestion. This will take some undoing of the things in this PR.

File: tests/bdd/features/ha/nexus-switchover/nexus_switchover.feature (hunk 6 above)
Comment (Abhinandan-Purkait) [reply tests/bdd/features/ha/nexus-switchover/nexus_switchover.feature]: Done.

File: control-plane/grpc/proto/v1/volume/volume.proto (hunk 5 above)
Comment (Abhinandan-Purkait) [reply control-plane/grpc/proto/v1/volume/volume.proto]: > by shutting down the existing target first

Should we add this? This step will be taken care of by the cluster agent, so would not be a part of publish call

Comment (tiagolobocastro) [reply control-plane/grpc/proto/v1/volume/volume.proto]: Isn't shutdown done by the core agent during republish?

Comment (Abhinandan-Purkait) [reply control-plane/grpc/proto/v1/volume/volume.proto]: I think the cluster agent does that. It shut downs the nexus and then make a republish call.

Comment (tiagolobocastro) [reply control-plane/grpc/proto/v1/volume/volume.proto]: I'm not sure that sounds right, the cluster agent does not own the nexus, it doesn't even know its uuid, only it's name.

Comment (Abhinandan-Purkait) [reply control-plane/grpc/proto/v1/volume/volume.proto]: Can it not get it from the volume spec? I was told the cluster agent drives all the steps.

**Comment by Abhinandan-Purkait** [comment]
bors try

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 7]
```diff
@@ -376,6 +376,93 @@ impl ResourcePublishing for OperationGuardArc<VolumeSpec> {
...
+
+        let nexus_node = get_volume_target_node(registry, &state, request, true).await?;
+        let newer_nexus_id = NexusId::new();
+
+        let operation =
+            VolumeOperation::Republish((nexus_node.clone(), newer_nexus_id.clone(), request.share));
+        let spec_clone = self.start_update(registry, &state, operation).await?;
+
+        let volume_target = spec_clone.target.as_ref().expect("already validated");
+        let mut older_nexus = specs.nexus(volume_target.nexus()).await?;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]: cannot use `?` here, must use validate_update_step ?

File: common/src/types/v0/transport/nexus.rs [hunk 8]
```diff
@@ -374,6 +378,14 @@ impl DestroyNexus {
     }
 }
 
+/// Shutdown Nexus Request.
+#[derive(Serialize, Deserialize, Default, Debug, Clone, Eq, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct ShutdownNexus {
+    /// The uuid of the nexus.
+    pub uuid: NexusId,
```
Comment (tiagolobocastro) [inline common/src/types/v0/transport/nexus.rs]: we've got loads of pub which around these files for "historical reasons", going forwards let's user getter instead, eg:
`fn uuid(&self) -> &NexusId { ... }`

File: control-plane/agents/src/bin/core/controller/reconciler/volume/hot_spare.rs [hunk 9]
```diff
@@ -64,6 +64,7 @@ async fn hot_spare_reconcile(
             hot_spare_nexus_reconcile(&mut volume, &volume_state, context).await
         }
         VolumeStatus::Faulted => PollResult::Ok(PollerState::Idle),
+        VolumeStatus::Shutdown => PollResult::Ok(PollerState::Idle),
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/volume/hot_spare.rs:67]: I think it's time for the VolumeStatus to break out of the NexusStatus?

File: control-plane/agents/src/bin/core/nexus/operations.rs [hunk 10]
```diff
@@ -245,3 +247,39 @@ impl ResourceOffspring for Option<&mut OperationGuardArc<NexusSpec>> {
...
+    type RemoveShutdownTargets = ();
+    type Shutdown = ShutdownNexus;
+
+    async fn shutdown(
+        &mut self,
+        registry: &Registry,
+        request: &Self::Shutdown,
+    ) -> Result<(), SvcError> {
+        let node_id = &self.lock().clone().node;
+        let nexus_id = &self.lock().clone().uuid;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/nexus/operations.rs]:
```suggestion
        let node_id = self.as_ref().node.clone();
        let nexus_id = self.uuid();
```

File: control-plane/agents/src/bin/core/nexus/operations.rs [hunk 11]
```diff
@@ -245,3 +247,39 @@ impl ResourceOffspring for Option<&mut OperationGuardArc<NexusSpec>> {
...
+        let node = registry.get_node_wrapper(node_id).await?;
+        let nexus_state = match node.nexus(nexus_id).await {
+            None => (&(self.lock().clone())).into(),
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/nexus/operations.rs]:
```suggestion
            None => self.as_ref().into(),
```

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 12]
```diff
@@ -376,6 +376,93 @@ impl ResourcePublishing for OperationGuardArc<VolumeSpec> {
...
+        let older_nexus_id = older_nexus.lock().clone().uuid;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]:
```suggestion
        let older_nexus_id = older_nexus.as_ref().uuid;
```

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 13]
```diff
@@ -376,6 +376,93 @@ impl ResourcePublishing for OperationGuardArc<VolumeSpec> {
...
+        // shutdown the older nexus before newer nexus creation.
+        older_nexus
+            .shutdown(
+                registry,
+                &ShutdownNexus {
+                    uuid: older_nexus_id,
+                },
+            )
+            .await?;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs:417]: can't return here

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 14]
```diff
@@ -376,6 +376,93 @@ impl ResourcePublishing for OperationGuardArc<VolumeSpec> {
...
+        let (mut nexus, nexus_state) = self
+            .validate_update_step(registry, result, &spec_clone)
+            .await?;
+
+        let (volume_id, last_nexus_id) = {
+            let volume_spec = self.lock();
+            (volume_spec.uuid.clone(), volume_spec.last_nexus_id.clone())
+        };
+
+        // Share the Nexus if it was requested
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]: I think republish should always imply re-sharing, so make it a requirement?

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 15]
```diff
@@ -376,6 +376,93 @@ impl ResourcePublishing for OperationGuardArc<VolumeSpec> {
...
+        self.complete_update(registry, result, spec_clone).await?;
+
+        // If there was a previous nexus we should delete the persisted NexusInfo structure.
+        if let Some(nexus_id) = last_nexus_id {
+            ResourceSpecsLocked::delete_nexus_info(
+                &NexusInfoKey::new(&Some(volume_id), &nexus_id),
+                registry,
+            )
+            .await;
+        }
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]: I wonder if this needs to be different for the republish, now.
Should we check if the previous nexus updated its nexus info, and if it did we should re-republish or update the nexus with the new healthy children list?

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 16]
```diff
@@ -407,3 +494,45 @@ impl ResourceReplicas for OperationGuardArc<VolumeSpec> {
...
+        registry: &Registry,
+        request: &Self::RemoveShutdownTargets,
+    ) -> Result<(), SvcError> {
+        let shutdown_nexuses = registry
+            .specs()
+            .get_volume_shutdown_nexuses_guards(&request.uuid)
+            .await?;
+        let mut all_targets_destroyed = true;
+        for mut nexus in shutdown_nexuses {
+            let nexus_spec = nexus.lock().clone();
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]:
```suggestion
            let nexus_spec = nexus.as_ref();
```

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 17]
```diff
@@ -407,3 +494,45 @@ impl ResourceReplicas for OperationGuardArc<VolumeSpec> {
...
+            if nexus.destroy(registry, &destroy_req).await.is_err() {
+                // flag that all targets weren't destroyed
+                all_targets_destroyed = false;
+            }
+        }
+        if all_targets_destroyed {
+            Ok(())
+        } else {
+            Err(SvcError::Deleting {})
+        }
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]: I think here we might simply return success as long as the specs are marked as deleting? We'd need to ensure we updated etcd correctly.

File: control-plane/agents/src/bin/core/volume/service.rs [hunk 18]
```diff
@@ -198,6 +227,19 @@ impl Service {
...
+    #[tracing::instrument(level = "info", skip(self), err, fields(volume.uuid = %request.uuid))]
+    pub(super) async fn destroy_shutdown_target(
+        &self,
+        request: &DestroyShutdownTargets,
+    ) -> Result<(), SvcError> {
+        let mut volume = self.specs().volume(&request.uuid).await?;
+        volume
+            .remove_shutdown_targets(&self.registry, request)
+            .await?;
+        Ok(())
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/service.rs]: keeping with the current style of the file:
```suggestion
        volume
            .remove_shutdown_targets(&self.registry, request)
            .await
```

File: control-plane/agents/src/bin/core/volume/specs.rs [hunk 19]
```diff
@@ -394,6 +394,28 @@ impl ResourceSpecsLocked {
             .cloned()
             .collect()
     }
+
+    /// Get a list of protected NexusSpecs's which are associated with the given volume `id`
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/specs.rs:399]: nexus guards?

File: control-plane/agents/src/bin/core/volume/specs.rs [hunk 20]
```diff
@@ -394,6 +394,28 @@ impl ResourceSpecsLocked {
...
+    /// and are currently in shutdown state.
+    pub(crate) async fn get_volume_shutdown_nexuses_guards(
+        &self,
+        id: &VolumeId,
+    ) -> Result<Vec<OperationGuardArc<NexusSpec>>, SvcError> {
+        let mut shutdown_nexuses = Vec::new();
+        for nexus in self.read().nexuses.values() {
+            let nexus_spec = nexus.lock().clone();
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/specs.rs]: why not use iter like the example above?

File: control-plane/agents/src/bin/core/volume/specs.rs [hunk 21]
```diff
@@ -394,6 +394,28 @@ impl ResourceSpecsLocked {
...
+            if nexus_spec.name == id.to_string() && nexus_spec.is_shutdown() {
+                shutdown_nexuses.push(nexus_spec);
+            }
+        }
+        let mut shutdown_nexuses_guards = Vec::with_capacity(shutdown_nexuses.len());
+        for nexus in shutdown_nexuses {
+            let nexus_guard = self.nexus(&nexus.uuid).await?;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/specs.rs]: this will error out if we can't grab one of the nexuses, is this the right thing to do?
What about returning only the Resource<Nexus>, and then try to guard them one at a time, and if we fail to guard one of them skip and fail the shutdown call, but at least we've cleaned up some of them?

File: control-plane/agents/src/bin/core/volume/specs.rs [hunk 22]
```diff
@@ -1240,6 +1265,17 @@ impl SpecOperationsHelper for VolumeSpec {
                     }),
                 },
             },
+            VolumeOperation::Republish((_, _, protocol)) => match protocol {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/specs.rs:1267]: I think protocol should not be optional here.

File: control-plane/rest/openapi-specs/v0_api_spec.yaml [hunk 23]
```diff
@@ -1437,6 +1437,15 @@ paths:
...
             $ref: '#/components/schemas/VolumeShareProtocol'
+        - in: query
+          name: republish
+          x-api-scope: internal
+          description: |-
+            Allows republishing the volume on the node by shutting down the existing target first.
+          required: false
+          schema:
+            type: boolean
+            default: false
```
Comment (tiagolobocastro) [inline control-plane/rest/openapi-specs/v0_api_spec.yaml]: does a default here make sense if republish is already optional?

**Review by tiagolobocastro — COMMENTED** [review]
Since this is quite large, might want to consider splitting this into different logical commits

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 12 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/operations.rs]: should it be rather - > `let older_nexus_id = older_nexus.uuid().clone();` Otherwise its giving moved error.

File: control-plane/agents/src/bin/core/controller/reconciler/volume/hot_spare.rs (hunk 9 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/controller/reconciler/volume/hot_spare.rs:67]: Can we break out? We use the nexus status to set volume status, what will we set if Nexus status is shutdown?

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 17 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/operations.rs]: The specs would be marked as deleting as that is the actual destroy call that takes care of it. I have added this error in case cluster agent needs to do something with the error. Otherwise we can simply return success.

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 12 above)
Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/volume/operations.rs]: Right yes needs to be cloned because it's moved.

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 17 above)
Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/volume/operations.rs]: I think cluster agent is happy to have us retry on our reconcilers IIRC @mtzaurus ?
But to be able to do that I think we need to check what kind of errors we got, example if we could not reach etcd then we can't ignore the error because if we restart, we'll "forget" that we tried to do this and won't reconcile.

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 14 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/operations.rs]: Made it a requirement

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 17 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/operations.rs]: Done.

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 15 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/operations.rs]: Have added comments for now, but needs revisiting after final discussion.

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 24]
```diff
@@ -376,6 +379,96 @@ impl ResourcePublishing for OperationGuardArc<VolumeSpec> {
...
+    async fn republish(
+        &mut self,
+        registry: &Registry,
+        request: &Self::Republish,
+    ) -> Result<Self::PublishOutput, SvcError> {
+        let specs = registry.specs();
+        let state = registry.get_volume_state(&request.uuid).await?;
+
+        // If the volume is not published then it should issue publish call rather than republish.
+        if state.target.is_none() {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]: should be from the spec, not state? and if we get spec we can skip the expect below, example:
```rust
let target = match spec.target { Some(target) => target, None => return NotPublish... }
```

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 25]
```diff
@@ -407,3 +500,67 @@ impl ResourceReplicas for OperationGuardArc<VolumeSpec> {
...
+        registry: &Registry,
+        request: &Self::RemoveShutdownTargets,
+    ) -> Result<(), SvcError> {
+        let shutdown_nexuses = registry
+            .specs()
+            .get_volume_shutdown_nexuses(&request.uuid)
+            .await;
+
+        for mut nexus_res in shutdown_nexuses {
+            let nexus_spec = nexus_res.lock().deref().clone();
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]: Don't think we need to do this ahead of time.
Below we can simply do guard.as_ref().clone() ?

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 26]
```diff
@@ -407,3 +500,67 @@ impl ResourceReplicas for OperationGuardArc<VolumeSpec> {
...
+                        },
+                    }
+                }
+                Err(err) => {
+                    debug!(
+                        err = %err,
+                        nexus.id = nexus_res.uuid().to_string(),
+                        "Encountered error while destroying shutdown nexus, exiting.."
+                    );
+                    return Err(err);
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]: Why exit here, just try the next ones.
```suggestion
```

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 27]
```diff
@@ -407,3 +500,67 @@ impl ResourceReplicas for OperationGuardArc<VolumeSpec> {
...
+                            }
+                        },
+                    }
+                }
+                Err(err) => {
+                    debug!(
+                        err = %err,
+                        nexus.id = nexus_res.uuid().to_string(),
+                        "Encountered error while destroying shutdown nexus, exiting.."
+                    );
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]:
```suggestion
                    debug!(
                        error = %err,
                        nexus.uuid = %nexus_res.uuid(),
                        "Failed to guard one of the destroying shutdown nexus"
                    );
```

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 28]
```diff
@@ -407,3 +500,67 @@ impl ResourceReplicas for OperationGuardArc<VolumeSpec> {
...
+                            .await;
+                        }
+                        Err(err) => match err {
+                            SvcError::Store { .. } => return Err(err),
+                            SvcError::StoreSave { .. } => return Err(err),
+                            _ => {
+                                debug!(
+                                    err = %err,
+                                    nexus.id = destroy_req.uuid.to_string(),
+                                    "Encountered error while destroying shutdown nexus"
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs]:
```suggestion
                                debug!(
                                    error = %err,
                                    nexus.uuid = %destroy_req.uuid,
                                    "Encountered error while destroying shutdown nexus"
```

File: control-plane/agents/src/bin/core/volume/specs.rs [hunk 29]
```diff
@@ -394,6 +395,24 @@ impl ResourceSpecsLocked {
...
+    /// and are currently in shutdown state.
+    pub(crate) async fn get_volume_shutdown_nexuses(
+        &self,
+        id: &VolumeId,
+    ) -> Vec<ResourceMutex<NexusSpec>> {
+        self.read()
+            .nexuses
+            .values()
+            .filter(|nexus| {
+                let nexus_spec = nexus.lock().clone();
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/specs.rs]: don't need to clone?

File: control-plane/agents/src/bin/core/volume/specs.rs [hunk 30]
```diff
@@ -394,6 +395,24 @@ impl ResourceSpecsLocked {
...
+                nexus_spec.name == id.to_string() && nexus_spec.is_shutdown()
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/specs.rs]: `nexus_spec.name == id.as_str()`

File: control-plane/grpc/src/operations/volume/traits.rs [hunk 31]
```diff
@@ -993,6 +1020,16 @@ impl From<&dyn PublishVolumeInfo> for PublishVolume {
...
 
+impl From<&dyn PublishVolumeInfo> for RepublishVolume {
+    fn from(data: &dyn PublishVolumeInfo) -> Self {
+        Self {
+            uuid: data.uuid(),
+            target_node: data.target_node(),
+            share: data.share().unwrap(),
+        }
+    }
+}
```
Comment (tiagolobocastro) [inline control-plane/grpc/src/operations/volume/traits.rs]: Is this needed?
and also `impl PublishVolumeInfo for RepublishVolume {` ?

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 32]
```diff
@@ -376,6 +379,96 @@ impl ResourcePublishing for OperationGuardArc<VolumeSpec> {
...
+            .validate_update_step(registry, result, &spec_clone)
+            .await?;
+
+        let older_nexus_id = older_nexus.uuid().clone();
+        // shutdown the older nexus before newer nexus creation.
+        let result = older_nexus
+            .shutdown(registry, &ShutdownNexus::new(older_nexus_id))
+            .await;
+        self.validate_update_step(registry, result, &spec_clone)
+            .await?;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs:417]: What happens if the node is offline and we can't shutdown? How should we handle this?
We can leave it for another PR? wdyt?

**Review by tiagolobocastro — COMMENTED** [review]
What happens if the node is offline and we can't shutdown? How should we handle this?
We can leave it for another PR, wdyt?

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 32 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/operations.rs:417]: Yes, right, this will error out if node goes offline, we need some way to handle this. Maybe we should allow grpc server not implemented errors and use the reservation keys to handle that scenario, and for any other error we error out?

File: control-plane/grpc/src/operations/volume/traits.rs (hunk 31 above)
Comment (Abhinandan-Purkait) [reply control-plane/grpc/src/operations/volume/traits.rs]: Yes, this is needed since we have seperated the structures for PublishVolume and RepublishVolume.

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 24 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/operations.rs]: Done.

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 32 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/operations.rs:417]: For now have added a comment. Needs revisiting again.

**Review by tiagolobocastro — APPROVED** [review]

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors merge
